src/common/valtext.cpp

```diff
@@ -332,6 +312,56 @@ void wxTextValidator::OnChar(wxKeyEvent& event)
     event.Skip(false);
 }
 
+bool wxTextValidator::IsValid(const wxUniChar& c) const
+{
+    if ( !m_validatorStyle ) // no filtering if HasFlag(wxFILTER_NONE)
+        return true;
+
+    if ( HasFlag(wxFILTER_SPACE) && wxIsspace(c) )
+        return true;
+
+    // Notice that if the same char is mistakenly present in both the
+    // include char list and exclude char list than the char will be
+    // excluded and a warning is given in debug build.
+
+    if ( IsCharExcluded(c) )
+        return false;
+
+    if ( IsCharIncluded(c) )
+        return true;
+
+    static const long mask = wxFILTER_ASCII|wxFILTER_ALPHA|wxFILTER_NUMERIC|
+                      wxFILTER_ALPHANUMERIC|wxFILTER_DIGITS|wxFILTER_XDIGITS;
+    
+    long flags = m_validatorStyle & mask;
+
+    if ( !flags )
+        // accept whatever this character is.
+        return true;
+
+    // Turn off the wxFILTER_XXX if the corresponding check failed...
+
+    if ( HasFlag(wxFILTER_ASCII) && !c.IsAscii() )
+        flags ^= wxFILTER_ASCII;
```

Comment:
Use of XOR seems like an obfuscation to me because it's always used to turn off the corresponding flag, not toggle it, so writing `flags &= ~wxFILTER_XXX` would be more clear.

Also, why not just return false here directly? A non-ASCII character won't be accepted in any case. In fact, we don't even need to test for `wxFILTER_ASCII`... I.e.
```
if ( !c.IsAscii() )
     return false;
```
would be correct AFAICS.

Reply:
@vadz  how about using this idiom:
```
bool success = true;
do
{
    if ( HasFlag(wxFILTER_ASCII) && !c.IsAscii() )
        break;
    if ( HasFlag(wxFILTER_NUMERIC) && !wxIsNumeric(c) )
        break;
    if ( HasFlag(wxFILTER_ALPHANUMERIC) && !wxIsalnum(c) )
        break;
    if ( HasFlag(wxFILTER_ALPHA) && !wxIsalpha(c) )
        break;
    if ( HasFlag(wxFILTER_DIGITS) && !wxIsdigit(c) )
        break;
    if ( HasFlag(wxFILTER_XDIGITS) && !wxIsxdigit(c) )
        break;

    // unknown char class! set success to false to break outside this loop
    // and indicate failure.
    success = false;

} while (success);

return success;

```

Reply:
Err, this doesn't work. If you break during the first iteration, the function returns true...

Reply:
Oops! i mean this:
```
bool success = true;
do
{
    if ( HasFlag(wxFILTER_ASCII) && c.IsAscii() )
        break;
    if ( HasFlag(wxFILTER_NUMERIC) && wxIsNumeric(c) )
        break;
    if ( HasFlag(wxFILTER_ALPHANUMERIC) && wxIsalnum(c) )
        break;
    if ( HasFlag(wxFILTER_ALPHA) && wxIsalpha(c) )
        break;
    if ( HasFlag(wxFILTER_DIGITS) && wxIsdigit(c) )
        break;
    if ( HasFlag(wxFILTER_XDIGITS) && wxIsxdigit(c) )
        break;

    // unknown char class! set success to false to break outside this loop
    // and indicate failure.
    success = false;

} while (success);

return success;

```

Reply:
What is the loop for? AFAICS it can't be executed more than once...

Reply:
sorry for wasting your precious time on this, i feel like dumb and i just ended up implementing it like this:
```
bool wxTextValidator::IsValidChar(const wxUniChar& c) const
{
    if ( !m_validatorStyle ) // no filtering if HasFlag(wxFILTER_NONE)
        return true;

    // Notice that if the same char is mistakenly present in both the
    // include char list and exclude char list than the char will be
    // excluded and a warning is given in debug build.

    if ( IsCharExcluded(c) )
        return false;

    if ( IsCharIncluded(c) )
        return true;

    if ( HasFlag(wxFILTER_SPACE) && wxIsspace(c) )
        return true;
    if ( HasFlag(wxFILTER_ASCII) && c.IsAscii() )
        return true;
    if ( HasFlag(wxFILTER_NUMERIC) && wxIsNumeric(c) )
        return true;
    if ( HasFlag(wxFILTER_ALPHANUMERIC) && wxIsalnum(c) )
        return true;
    if ( HasFlag(wxFILTER_ALPHA) && wxIsalpha(c) )
        return true;
    if ( HasFlag(wxFILTER_DIGITS) && wxIsdigit(c) )
        return true;
    if ( HasFlag(wxFILTER_XDIGITS) && wxIsxdigit(c) )
        return true;

    // unknown char class! return false to indicate failure.
    return false;
}

```
i just renamed `IsValid()` to `IsValidChar()`
`wxFILTER_SPACE` is a convenience flag added if one want to allow spaces to be entered alongside alphas or digits or both (which is a common scenario in text validators) in a simple way without the need to call `SetCharIncludes(" ")` just for this.

Reply:
Yes, this is good and looks much simpler and obvious than the original version, thanks!